File: backend/app/com/skewwhiffy/auraltester/clefs/ClefLowerLedgerNotes.scala
package com.skewwhiffy.auraltester.clefs

import com.skewwhiffy.auraltester.notes.AbsoluteNote

case class ClefLowerLedgerNotes(clef: Clef) extends ClefNotes {
  def notes: List[AbsoluteNote] = {
    val highestNote = clef.lowLedgerNote.upOne
    Range(1, 6)
      .foldLeft(List(highestNote))((soFar, _) => soFar :+ soFar.last.downOne)
      .map(it => it.withNoteName)
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/model/abc/ClefLowerLedgerNotes.kt
package com.skewwhiffy.notation.model.abc

import com.skewwhiffy.notation.model.note.AbsoluteNote
import com.skewwhiffy.notation.model.clef.Clef

data class ClefLowerLedgerNotes(private val clef: Clef) : ClefNotes {
  override val notes: List<AbsoluteNote> = (1..5)
    .fold(listOf(clef.lowLedgerNote.upOne)) { soFar, _ -> soFar + soFar.last().downOne }
    .map { it.withNoteName }
}